pallets/gated-marketplace/src/functions.rs

```diff
@@ -150,9 +151,328 @@ impl<T: Config> Pallet<T> {
+        pallet_uniques::Pallet::<T>::do_transfer(collection_id, item_id, buy_offer_creator.clone(), |_, _|{
+            Ok(())
+        })?;
+
+        //update offer status from all marketplaces
+        Self::update_offers_status(buy_offer_creator.clone(), collection_id, item_id, marketplace_id)?;
+
+        //remove all the offers associated with the item
+        Self::delete_all_offers_for_this_item(collection_id, item_id )?;
+
+        Self::deposit_event(Event::OfferWasAccepted(offer_id, buy_offer_creator));
+        Ok(())
+    }
+
+    pub fn do_duplicate_offer(authority: T::AccountId, offer_id: [u8;32], marketplace_id: [u8;32], collection_id: T::CollectionId, item_id: T::ItemId, modified_price: BalanceOf<T>) -> DispatchResult{
+        //ensure new marketplace_id exits
+        ensure!(<Marketplaces<T>>::contains_key(marketplace_id), Error::<T>::MarketplaceNotFound);
+
+        //ensure that the offer_id exists
+        ensure!(<OffersInfo<T>>::contains_key(offer_id), Error::<T>::OfferNotFound);
+
+        //ensure the offer_id exists in OffersByItem
+        Self::does_exist_offer_id_for_this_item(collection_id, item_id, offer_id)?;
+
+        //get the offer data
+        let mut copy_offer_data = <OffersInfo<T>>::get(offer_id).ok_or(Error::<T>::OfferNotFound)?;
+
+        //modify the offer data
+        //by know we only allow to modify its price by the user
+        //we modify its marketplace_id because the offer is duplicated to another marketplace
+        copy_offer_data.price = modified_price;
+        copy_offer_data.marketplace_id = marketplace_id;
+
+        //generate a new offer_id
+        let new_offer_id = (marketplace_id, authority.clone(), collection_id, copy_offer_data.creation_date, copy_offer_data.expiration_date).using_encoded(blake2_256);
+
+        //insert in OffersInfo
+        // validate new offer_id does not exists
+        ensure!(!<OffersInfo<T>>::contains_key(new_offer_id), Error::<T>::OfferAlreadyExists);
+        <OffersInfo<T>>::insert(new_offer_id, copy_offer_data);
+
+        //insert in OffersByMarketplace
+        <OffersByMarketplace<T>>::try_mutate(marketplace_id, |offer| {
+            offer.try_push(new_offer_id)
+        }).map_err(|_| Error::<T>::OfferStorageError)?; 
+        
+        //insert in OffersByAccount
+        <OffersByAccount<T>>::try_mutate(authority.clone(), |offer| {
+            offer.try_push(new_offer_id)
+        }).map_err(|_| Error::<T>::OfferStorageError)?;
+        
+        //add the new offer_id to OffersByItem
+        <OffersByItem<T>>::try_mutate(collection_id, item_id, |offers| {
+            offers.try_push(new_offer_id)
+        }).map_err(|_| Error::<T>::OfferStorageError)?;
+
+        Self::deposit_event(Event::OfferDuplicated(new_offer_id, marketplace_id));
+
+        Ok(())
+    }
+
+    pub fn do_remove_offer(authority: T::AccountId, offer_id: [u8;32], marketplace_id: [u8;32], collection_id: T::CollectionId, item_id: T::ItemId, ) -> DispatchResult {
```

Comment:
I think the only paramaters should be authority and offer_id, the other parameters we can get from the offer details

Reply:
You're right, I updated to version 5. Done.